src/core/load_balancing/rls/rls.cc

```diff
@@ -1197,9 +1209,16 @@ void RlsLb::Cache::Entry::Orphan() {
   GRPC_TRACE_LOG(rls_lb, INFO)
       << "[rlslb " << lb_policy_.get() << "] cache entry=" << this << " "
       << lru_iterator_->ToString() << ": cache entry evicted";
-  is_shutdown_ = true;
-  lb_policy_->cache_.lru_list_.erase(lru_iterator_);
-  lru_iterator_ = lb_policy_->cache_.lru_list_.end();  // Just in case.
+  std::vector<RefCountedPtr<ChildPolicyWrapper>>
+      child_policy_wrappers_to_delete;
+  {
+    MutexLock lock(&lb_policy_->mu_);
```

Comment:
It seems a little inefficient to wait until after releasing the lock to orphan the entry but then reacquire the lock when actually orphaning it.  Instead, I suggest that right before we orphan the entry, we move `child_policy_wrappers_` out from inside of it, and then we orphan the entry while still holding the lock.  That way, the only thing we need to unref after releasing the lock are the `ChildPolicyWrapper`s themselves, which won't require re-acquiring the lock.

Reply:
Done